server/src/main/java/org/elasticsearch/search/query/QuerySearchResult.java

```diff
@@ -405,4 +422,37 @@ public TotalHits getTotalHits() {
     public float getMaxScore() {
         return maxScore;
     }
+
+    @Override
+    public void incRef() {
+        refCounted.incRef();
+    }
+
+    @Override
+    public boolean tryIncRef() {
+        return refCounted.tryIncRef();
+    }
+
+    @Override
+    public boolean decRef() {
+        return refCounted.decRef();
+    }
+
+    @Override
+    public boolean hasReferences() {
+        return refCounted.hasReferences();
+    }
+
+    // todo: remove this and the logger before merging
+    @Override
+    protected void finalize() throws Throwable {
+        if (Assertions.ENABLED) {
+            if (refCounted.hasReferences()) {
```

Comment:
In order to make this not fire, we would need to change how `SearchContext` creates these results or add inc/decRef there. I opt'ed not to do this, since this more targeted fix should work out fine, but would value opinions on that.